xdsl/interpreter.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from io import StringIO
+from typing import Any, Callable, Generator, TypeVar
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import SSAValue, Operation
+
+_OperationInvT = TypeVar('_OperationInvT', bound=Operation)
+
+
+class IntepretationError(Exception):
+    pass
+
+
+@dataclass
+class Functions:
+    functions: dict[type[Operation],
+                    Callable[[Intepreter, Operation, tuple[Any, ...]],
+                             tuple[Any, ...]]] = field(default_factory=dict)
+
+    def register_op(
+        self, op_type: type[_OperationInvT],
+        func: Callable[[Intepreter, _OperationInvT, tuple[Any, ...]],
+                       tuple[Any, ...]]):
+        self.functions[op_type] = func  # type: ignore
+
+    def op_types(self) -> set[type[Operation]]:
+        return set(self.functions.keys())
+
+    def register(
+        self, op_type: type[_OperationInvT]
+    ) -> Callable[[
+            Callable[[Intepreter, _OperationInvT, tuple[Any, ...]], tuple[Any,
+                                                                          ...]]
+    ], Callable[[Intepreter, _OperationInvT, tuple[Any, ...]], tuple[Any,
+                                                                     ...]]]:
+
+        def wrapper(
+            func: Callable[[Intepreter, _OperationInvT, tuple[Any, ...]],
+                           tuple[Any, ...]]):
+            self.register_op(op_type, func)
+            return func
+
+        return wrapper
+
+    def run(self, interpreter: Intepreter, op: Operation,
+            args: tuple[Any, ...]) -> tuple[Any, ...]:
+        return self.functions[type(op)](interpreter, op, args)
+
+    def register_from(self, other: Functions):
+        '''If there are duplicate definitions, the `other` will override `self`'''
+        self.functions.update(other.functions)
+
+
+@dataclass
+class IntepretationEnv:
+    name: str = field(default="unknown")
```

Comment:
Is this even needed? This is only set for the root, right? For the root we already know it's the root as it has no parent, right?

Reply:
no, I actually add the function name in there, it's quite useful for debugging even in the current hacky form

Reply:
Ah, okay did not see :+1: